ingest/validation/validate_metadata.py

```diff
@@ -127,6 +129,239 @@ def create_parser():
+
+
+def backoff_handler(details):
+    """Handler function to log backoff attempts when querying OLS
+    """
+    info_logger.debug(
+        "Backing off {wait:0.1f} seconds after {tries} tries "
+        "calling function {target} with args {args} and kwargs "
+        "{kwargs}".format(**details),
+        extra={'study_id': None, 'duration': None},
+    )
+
+
+# contains methods for looking up terms in various ontologies,
+# as well as caching results of previous queries to speed up performance
+class OntologyRetriever:
+    cached_ontologies = {}
+    cached_terms = {}
+
+    def retrieve_ontology_term_label(self, term, property_name, convention):
+        """Retrieve an individual term label from an ontology
+        returns JSON payload of ontology, or None if unsuccessful
+        Will store any retrieved terms for faster validation of downstream terms
+
+        throws ValueError if the term does not exist or is malformatted
+        """
+
+        # initialize cached terms for this property if we haven't seen it yet
+        if property_name not in self.cached_terms:
+            self.cached_terms[property_name] = {}
+
+        if term not in self.cached_terms[property_name]:
+            ontology_urls = get_urls_for_property(convention, property_name)
+            self.cached_terms[property_name][
+                term
+            ] = self.retrieve_ontology_term_label_remote(
+                term, property_name, ontology_urls
+            )
+
+        return self.cached_terms[property_name][term]
+
+    def retrieve_ontology_term_label_remote(self, term, property_name, ontology_urls):
+        """Look up official ontology term from an id, always checking against the remote
+        """
+        if property_name == 'organ_region':
+            return self.retrieve_mouse_brain_term(term, property_name)
+        else:
+            return self.retrieve_ols_term(ontology_urls, term, property_name)
+
+    # Attach exponential backoff to external HTTP requests
+    @backoff.on_exception(
+        backoff.expo,
+        requests.exceptions.RequestException,
+        max_time=MAX_HTTP_REQUEST_TIME,
+        max_tries=MAX_HTTP_ATTEMPTS,
+        on_backoff=backoff_handler,
+        logger="info_logger",
+    )
+    def retrieve_ols_term(self, ontology_urls, term, property_name):
+        """Retrieve an individual term from an ontology
+        returns JSON payload of ontology, or None if unsuccessful
+        Will store any retrieved ontologies for faster validation of downstream terms
+        """
+        OLS_BASE_URL = 'https://www.ebi.ac.uk/ols/api/ontologies/'
+        # separate ontology shortname from term ID number
+        # valid separators are underscore and colon (used by HCA)
+        try:
+            ontology_shortname, term_id = re.split('[_:]', term)
+        except (ValueError, TypeError):
+            raise ValueError(
+                f'{property_name}: No match found in EBI OLS for provided ontology ID: {term}'
```

Comment:
This error seems misleading since we haven't made the request to OLS yet.

Reply:
Thanks for catching this, Jon. I suspect Devon used that error message because it allowed an existing test to pass. I've updated the error message to be more appropriate and updated the test's reference file to reflect the new error messaging.